tests/test_mpls.py

```diff
@@ -0,0 +1,598 @@
+
+        # get neighbor and arp entry
+        if resolve:
+            dvs.servers[0].runcmd("ping -c 1 10.0.0.3")
+            dvs.servers[2].runcmd("ping -c 1 10.0.0.3")
+
+    def teardown_mpls(self, dvs):
+        # remove ip address
+        self.remove_ip_address("Ethernet0", "10.0.0.0/31")
+        self.remove_ip_address("Ethernet4", "10.0.0.2/31")
+        self.remove_ip_address("Ethernet8", "10.0.0.4/31")
+
+        # remove mpls interface
+        self.remove_mpls_intf("Ethernet0")
+        self.remove_mpls_intf("Ethernet4")
+        self.remove_mpls_intf("Ethernet8")
+
+        self.set_admin_status("Ethernet0", "down")
+        self.set_admin_status("Ethernet4", "down")
+        self.set_admin_status("Ethernet8", "down")
+
+        # remove ip address and default route
+        dvs.servers[0].runcmd("ip route del default dev eth0")
+        dvs.servers[0].runcmd("ip address del 10.0.0.1/31 dev eth0")
+
+        dvs.servers[1].runcmd("ip route del default dev eth0")
+        dvs.servers[1].runcmd("ip address del 10.0.0.3/31 dev eth0")
+
+        dvs.servers[2].runcmd("ip route del default dev eth0")
+        dvs.servers[2].runcmd("ip address del 10.0.0.5/31 dev eth0")
+
+        dvs.runcmd("rmmod mpls_iptunnel")
+        dvs.runcmd("rmmod mpls_router")
+
+class TestMplsRoute(TestMplsBase):
+    """ Functionality tests for mpls """
+    def test_RouteAddRemoveIpRoutePush(self, dvs, testlog):
+        self.setup_mpls(dvs, True)
+
+        # add route entry
+        prefix = "2.2.2.0/24"
+        if self.mpls_appdb_mode():
+            fieldValues = {"nexthop": "10.0.0.1", "ifname": "Ethernet0", "mpls_nh": "push+201"}
+            self.create_route_entry(prefix, fieldValues)
+        else:
+            # dvs.runcmd("ip route add 2.2.2.0/24 encap mpls 201 via inet 10.0.0.1 dev Ethernet0")
+            dvs.runcmd("vtysh -c \"configure terminal\" -c \"ip route 2.2.2.0/24 10.0.0.1 Ethernet0 label 201\"")
+
+        # check application database
+        self.pdb.wait_for_entry("ROUTE_TABLE", prefix)
+        time.sleep(1)
+
+        # check ASIC route database
+        self.check_route_entries(True, [prefix])
+        self.check_route_nexthop(prefix, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_PUSH", "1:201")
+
+        # remove route entry
+        if self.mpls_appdb_mode():
+            self.remove_route_entry(prefix)
+        else:
+            # dvs.runcmd("ip route del 2.2.2.0/24 encap mpls 201 via inet 10.0.0.1 dev Ethernet0")
+            dvs.runcmd("vtysh -c \"configure terminal\" -c \"no ip route 2.2.2.0/24 10.0.0.1 Ethernet0 label 201\"")
+
+        # check application database
+        self.pdb.wait_for_deleted_entry("ROUTE_TABLE", prefix)
+        time.sleep(1)
+
+        # check ASIC route database
+        self.check_route_entries(False, [prefix])
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_PUSH", "1:201")
```

Comment:
assert based on the return value for all check_nexthop calls?

Reply:
@kperumalbfn check_nexthop() doesn't return anything.  It will assert inside check_nexthop() if something isn't as expected.